# தரவுத் தயாரிப்பு

[மூல நோட்புக் ஆதாரம் *தரவியல்: தரவியல் Python மற்றும் Machine Learning Studio கற்கும் இயந்திர கற்றலுக்கான அறிமுகம் - லீ ஸ்டாட்*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## `DataFrame` தகவல்களை ஆராய்வது

> **கற்கும் இலக்கு:** இந்த துணை பிரிவின் முடிவில், pandas DataFrames-ல் சேமிக்கப்பட்ட தரவின் பொதுவான தகவல்களை கண்டறிய நீங்கள் வசதியாக இருக்க வேண்டும்.

நீங்கள் உங்கள் தரவுகளை pandas-க்கு ஏற்றியவுடன், அது `DataFrame` வடிவத்தில் இருக்கும் வாய்ப்பு அதிகம். ஆனால், உங்கள் `DataFrame`-ல் உள்ள தரவுத்தொகுப்பில் 60,000 வரிசைகள் மற்றும் 400 நெடுவரிசைகள் இருந்தால், நீங்கள் எதிலிருந்து தொடங்குவது? அதிர்ஷ்டவசமாக, pandas ஒரு `DataFrame` பற்றிய மொத்த தகவல்களை விரைவாகப் பார்க்கவும், முதல் சில மற்றும் கடைசி சில வரிசைகளைப் பார்க்கவும் வசதியான கருவிகளை வழங்குகிறது.

இந்த செயல்பாட்டை ஆராய்வதற்காக, Python scikit-learn நூலகத்தை இறக்குமதி செய்து, ஒவ்வொரு தரவியல் விஞ்ஞானியும் நூற்றுக்கணக்கான முறை பார்த்திருக்கும் ஒரு பிரபலமான தரவுத்தொகுப்பைப் பயன்படுத்துவோம்: பிரிட்டிஷ் உயிரியல் விஞ்ஞானி ரொனால்ட் ஃபிஷரின் *Iris* தரவுத்தொகுப்பு, 1936-ல் அவர் எழுதிய "Taxonomic பிரச்சினைகளில் பல அளவீடுகளைப் பயன்படுத்துதல்" என்ற கட்டுரையில் பயன்படுத்தப்பட்டது:


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
நாம் `iris_df` என்ற மாறியில் Iris Dataset-ஐ ஏற்றியுள்ளோம். தரவுகளை ஆராய்வதற்கு முன், எத்தனை தரவுப் புள்ளிகள் உள்ளன மற்றும் தரவுத்தொகுப்பின் மொத்த அளவு என்ன என்பதை அறிந்துகொள்வது பயனுள்ளதாக இருக்கும். நாம் கையாளும் தரவின் அளவைப் பார்ப்பது முக்கியம்.


In [2]:
iris_df.shape

(150, 4)

அப்படியானால், நாங்கள் 150 வரிசைகள் மற்றும் 4 நெடுவரிசைகளைக் கொண்ட தரவுகளை கையாளுகிறோம். ஒவ்வொரு வரிசையும் ஒரு தரவுப் புள்ளியை பிரதிநிதித்துவப்படுத்துகிறது, மேலும் ஒவ்வொரு நெடுவரிசையும் தரவுச் சட்டகத்துடன் தொடர்புடைய ஒரு தனித்தன்மையை பிரதிநிதித்துவப்படுத்துகிறது. அடிப்படையாக, 150 தரவுப் புள்ளிகள் உள்ளன, ஒவ்வொன்றும் 4 தனித்தன்மைகளை கொண்டுள்ளது.

`shape` என்பது இங்கு dataframe இன் ஒரு பண்பாகும், ஒரு செயல்பாடு அல்ல, அதனால் இது ஒரு ஜோடி வட்டக் கோடுகளால் முடிவடைவதில்லை.


### `DataFrame.columns`
இப்போது நான்கு களங்களின் தரவுக்குள் செல்லலாம். ஒவ்வொன்றும் என்னைச் சரியாக பிரதிநிதித்துவப்படுத்துகிறது? `columns` பண்புக்கூறு dataframe இல் உள்ள களங்களின் பெயரை எங்களுக்கு வழங்கும்.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

நாம் காணக்கூடியதுபோல், நான்கு (4) sütunlar உள்ளன. `columns` பண்புக்கூறு sütunların பெயர்களைச் சொல்கிறது மற்றும் அடிப்படையில் வேறொன்றும் இல்லை. இந்த பண்புக்கூறு ஒரு dataset உள்ள அம்சங்களை அடையாளம் காண விரும்பும்போது முக்கியத்துவம் பெறுகிறது.


### `DataFrame.info`
`shape` பண்புக்கூறு மூலம் தரப்பட்ட தரவின் அளவு மற்றும் `columns` பண்புக்கூறு மூலம் தரப்பட்ட அம்சங்கள் அல்லது நெடுக்களின் பெயர்கள் dataset பற்றி எதையாவது கூறுகின்றன. இப்போது, நாம் dataset-ஐ மேலும் ஆழமாக ஆராய விரும்புகிறோம். `DataFrame.info()` செயல்பாடு இதற்காக மிகவும் பயனுள்ளதாக உள்ளது.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


இங்கிருந்து, சில கவனிக்கத்தக்க விஷயங்களை பார்க்கலாம்:  
1. ஒவ்வொரு நெடுவரிசையின் தரவுத்தரம்: இந்த தரவுத்தொகுப்பில், அனைத்து தரவுகளும் 64-பிட் மிதவை புள்ளி எண்களாக சேமிக்கப்பட்டுள்ளன.  
2. Non-Null மதிப்புகளின் எண்ணிக்கை: Null மதிப்புகளை கையாளுவது தரவுத் தயாரிப்பில் ஒரு முக்கியமான படியாகும். இது பின்னர் நோட்புக்கில் கையாளப்படும்.  


### DataFrame.describe()
நாம் எங்கள் தரவுத்தொகுப்பில் நிறைய எண் தரவுகளை வைத்திருக்கிறோம் என்று கூறுங்கள். சராசரி, நடுநிலை, குவார்டைல்கள் போன்ற ஒற்றைமாறி புள்ளியியல் கணக்குகளை ஒவ்வொரு நெடுவரிசையிலும் தனித்தனியாக செய்யலாம். `DataFrame.describe()` செயல்பாடு தரவுத்தொகுப்பின் எண் நெடுவரிசைகளின் புள்ளியியல் சுருக்கத்தை எங்களுக்கு வழங்குகிறது.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

மேலே உள்ள வெளியீடு ஒவ்வொரு நெடுவரிசையின் மொத்த தரவுப் புள்ளிகள், சராசரி, நிலையான சிதறல், குறைந்தபட்சம், கீழ் குவார்டைல்(25%), நடுக்கோடு(50%), மேல் குவார்டைல்(75%) மற்றும் அதிகபட்ச மதிப்பை காட்டுகிறது.


### `DataFrame.head`
மேலே உள்ள அனைத்து செயல்பாடுகள் மற்றும் பண்புகளுடன், தரவுத்தொகுப்பின் மேல் நிலைப் பார்வையை பெற்றுள்ளோம். எத்தனை தரவுப் புள்ளிகள் உள்ளன, எத்தனை அம்சங்கள் உள்ளன, ஒவ்வொரு அம்சத்தின் தரவுத் வகை மற்றும் ஒவ்வொரு அம்சத்திற்கும் non-null மதிப்புகளின் எண்ணிக்கை ஆகியவற்றை அறிந்துள்ளோம்.

இப்போது தரவுகளை நேரடியாகப் பார்க்க நேரம் வந்துள்ளது. எங்கள் `DataFrame`-இன் முதல் சில வரிகள் (முதல் சில தரவுப் புள்ளிகள்) எப்படி இருக்கின்றன என்பதை பார்ப்போம்:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

இங்கே உள்ள வெளியீட்டைப் பார்க்கும்போது, தரவுத்தொகுப்பின் ஐந்து(5) பதிவுகளை காணலாம். இடதுபுறத்தில் உள்ள குறியீட்டைப் பார்க்கும்போது, இவை முதல் ஐந்து வரிசைகள் என்பதை கண்டறிகிறோம்.


### பயிற்சி:

மேற்கண்ட எடுத்துக்காட்டில் இருந்து, இயல்பாக, `DataFrame.head` ஒரு `DataFrame`-இன் முதல் ஐந்து வரிசைகளை திருப்பி அளிக்கிறது என்பது தெளிவாகிறது. கீழே உள்ள குறியீட்டு செல்லில், ஐந்து வரிசைகளுக்கு மேல் காட்ட ஒரு வழியை கண்டுபிடிக்க முடியுமா?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
தரவைப் பார்க்கும் மற்றொரு வழி முடிவிலிருந்து (தொடக்கத்தின் பதிலாக) இருக்கலாம். `DataFrame.head`-இன் எதிர்மறை `DataFrame.tail` ஆகும், இது ஒரு `DataFrame`-இன் கடைசி ஐந்து வரிசைகளை திருப்பி அளிக்கிறது:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

நடப்பில், `DataFrame`-இன் முதல் சில வரிசைகள் அல்லது கடைசி சில வரிசைகளை எளிதாகப் பார்வையிட முடிவது மிகவும் பயனுள்ளதாக இருக்கும், குறிப்பாக ஒழுங்குபடுத்தப்பட்ட தரவுத்தொகுப்புகளில் வெளிப்படையான மதிப்புகளைத் தேடும்போது.

மேலே கொடுக்கப்பட்ட கோடுகளின் உதவியுடன் காட்டப்பட்ட அனைத்து செயல்பாடுகள் மற்றும் பண்புகள், தரவின் தோற்றத்தையும் உணர்வையும் பெற உதவுகின்றன.

> **முக்கியக் குறிப்புகள்:** ஒரு DataFrame-இல் உள்ள தகவலின் மெட்டாடேட்டா அல்லது அதில் உள்ள முதல் மற்றும் கடைசி சில மதிப்புகளைப் பார்ப்பதன் மூலம், நீங்கள் கையாளும் தரவின் அளவு, வடிவம் மற்றும் உள்ளடக்கத்தை உடனடியாகப் புரிந்துகொள்ள முடியும்.


### காணாமல் போன தரவுகள்
காணாமல் போன தரவுகளைப் பற்றி விரிவாகப் பார்ப்போம். சில காலங்களில், சில பத்திகளில் எந்த மதிப்பும் சேமிக்கப்படாதபோது காணாமல் போன தரவுகள் ஏற்படுகின்றன.

ஒரு உதாரணத்தை எடுத்துக்கொள்வோம்: ஒருவருக்கு அவரது/அவளது எடை குறித்து கவலை இருந்தால், அவர் ஒரு கணக்கெடுப்பில் எடை புலத்தை நிரப்பாமல் விடுவார். அப்போது, அந்த நபரின் எடை மதிப்பு காணாமல் போயிருக்கும்.

உலகளாவிய தரவுத்தொகுப்புகளில், பெரும்பாலான நேரங்களில், காணாமல் போன மதிப்புகள் ஏற்படுகின்றன.

**Pandas காணாமல் போன தரவுகளை எவ்வாறு கையாளுகிறது**

Pandas காணாமல் போன மதிப்புகளை இரண்டு வழிகளில் கையாளுகிறது. முதலில், நீங்கள் முந்தைய பிரிவுகளில் பார்த்த `NaN` (Not a Number) ஆகும். இது IEEE floating-point specification இன் ஒரு சிறப்பு மதிப்பாகும், மற்றும் இது காணாமல் போன floating-point மதிப்புகளை குறிக்க மட்டுமே பயன்படுத்தப்படுகிறது.

Floating-point மதிப்புகளைத் தவிர, காணாமல் போன மதிப்புகளுக்கு pandas Python இன் `None` பொருளைப் பயன்படுத்துகிறது. இரண்டு விதமான மதிப்புகளை சந்திக்க நேரிடும் போது இது குழப்பமாக இருக்கலாம் என்று தோன்றினாலும், இந்த வடிவமைப்பு தேர்வுக்கு ஒழுங்கான நிரலாக்க காரணங்கள் உள்ளன, மேலும் நடைமுறையில், இந்த வழி pandas-க்கு பெரும்பாலான சந்தர்ப்பங்களில் நல்ல சமநிலையை வழங்க உதவுகிறது. இதைத் தவிர, `None` மற்றும் `NaN` இரண்டிலும் சில கட்டுப்பாடுகள் உள்ளன, அவற்றை எவ்வாறு பயன்படுத்த வேண்டும் என்பதில் நீங்கள் கவனமாக இருக்க வேண்டும்.


### `None`: non-float மிஸ்ஸிங் டேட்டா
`None` என்பது Python-இல் இருந்து வந்ததால், இது NumPy மற்றும் pandas அடர்த்திகளில் `'object'` டேட்டா வகை அல்லாதவற்றில் பயன்படுத்த முடியாது. நினைவில் கொள்ளுங்கள், NumPy அடர்த்திகள் (மற்றும் pandas-இல் உள்ள டேட்டா அமைப்புகள்) ஒரே வகையான டேட்டாவை மட்டுமே கொண்டிருக்க முடியும். இதுவே பெரிய அளவிலான டேட்டா மற்றும் கணினி வேலைகளுக்கு அவற்றின் மிகப்பெரிய சக்தியை வழங்குகிறது, ஆனால் இது அவற்றின் நெகிழ்வுத்தன்மையை வரையறுக்கிறது. இவ்வகையான அடர்த்திகள் "குறைந்த பொதுவான பரிமாணம்" எனப்படும், அதாவது அடர்த்தியில் உள்ள அனைத்தையும் உள்ளடக்கக்கூடிய டேட்டா வகைக்கு மேம்படுத்தப்பட வேண்டும். `None` அடர்த்தியில் இருந்தால், நீங்கள் Python பொருட்களுடன் வேலை செய்கிறீர்கள் என்று பொருள்.

இதனை செயல்பாட்டில் காண, கீழே உள்ள எடுத்துக்காட்டு அடர்த்தியை கவனிக்கவும் (அதற்கான `dtype`-ஐ கவனிக்கவும்):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

உயர்த்தப்பட்ட தரவுத் வகைகளின் உண்மை இரண்டு பக்க விளைவுகளை கொண்டுள்ளது. முதலில், செயல்பாடுகள் தொகுக்கப்பட்ட NumPy குறியீட்டின் பதிலாக, மொழிபெயர்க்கப்பட்ட Python குறியீட்டின் மட்டத்தில் செயல்படுத்தப்படும். அடிப்படையாக, இதன் பொருள் `Series` அல்லது `DataFrames`-இல் `None` உள்ள செயல்பாடுகள் மெதுவாக இருக்கும். நீங்கள் இந்த செயல்திறன் குறைவைக் கவனிக்காமல் இருக்கலாம், ஆனால் பெரிய தரவுத்தொகுப்புகளுக்கு இது ஒரு பிரச்சனையாக மாறக்கூடும்.

இரண்டாவது பக்க விளைவு முதல் விளைவிலிருந்து தோன்றுகிறது. ஏனெனில் `None` அடிப்படையாக `Series` அல்லது `DataFrame`s-ஐ சாதாரண Python உலகிற்கு திரும்பத் தள்ளுகிறது, `sum()` அல்லது `min()` போன்ற NumPy/pandas தொகுப்புகளை `None` மதிப்புகளை கொண்ட வரிசைகளில் பயன்படுத்துவது பொதுவாக ஒரு பிழையை உருவாக்கும்:


In [10]:
example1.sum()

TypeError: ignored

**முக்கியக் குறிப்புகள்**: முழு எண்கள் மற்றும் `None` மதிப்புகளுக்கு இடையிலான கூட்டல் (மற்றும் பிற செயல்பாடுகள்) வரையறுக்கப்படவில்லை, இது அவற்றைக் கொண்ட தரவுத்தொகுப்புகளுடன் நீங்கள் செய்யக்கூடியவற்றை வரையறுக்க முடியும்.


### `NaN`: மிஸ்ஸிங் ஃப்ளோட் மதிப்புகள்

`None`-க்கு மாறாக, NumPy (அதனால் pandas) அதன் வேகமான, வெக்டரைச்ட் செயல்பாடுகள் மற்றும் ufuncs-க்கு `NaN`-ஐ ஆதரிக்கிறது. மோசமான செய்தி என்னவென்றால், `NaN`-இல் செய்யப்படும் எந்த கணித செயல்பாடும் எப்போதும் `NaN`-ஐ உருவாக்கும். உதாரணமாக:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

நல்ல செய்தி: `NaN` உள்ள வரிசைகளில் தொகுப்புகள் இயக்கப்படும் போது பிழைகள் தோன்றவில்லை. கெட்ட செய்தி: முடிவுகள் ஒரே மாதிரியான பயனுள்ளவை அல்ல:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### பயிற்சி:


In [11]:
# What happens if you add np.nan and None together?


நினைவில் கொள்ளவும்: `NaN` என்பது மிஸ்ஸிங் ப்ளோட்டிங்-பாயிண்ட் மதிப்புகளுக்காக மட்டுமே; முழு எண்கள், சரங்கள் அல்லது பூலியன்களுக்கு `NaN` சமமானது இல்லை.


### `NaN` மற்றும் `None`: pandas-இல் null மதிப்புகள்

`NaN` மற்றும் `None` சில அளவுக்கு வேறுபடக்கூடியதாக செயல்படினாலும், pandas அவற்றை பரஸ்பரம் மாற்றி நடத்துவதற்காக உருவாக்கப்பட்டுள்ளது. இதை விளக்கமாகப் புரிந்துகொள்ள, முழு எண்களின் ஒரு `Series`-ஐ கவனிக்கவும்:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### பயிற்சி:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


`Series` மற்றும் `DataFrame`களில் தரவின் ஒரே மாதிரியான தன்மையை உருவாக்கும் நோக்கில் தரவின் வகைகளை மேம்படுத்தும் செயல்முறையில், pandas `None` மற்றும் `NaN` ஆகியவற்றை காணாமல் போன மதிப்புகளுக்கு இடையே மாற்றுவதற்கு தயங்காது. இந்த வடிவமைப்பு அம்சத்தால், pandas-இல் `None` மற்றும் `NaN` ஆகியவற்றை "null" எனும் இரண்டு வெவ்வேறு வகைகளாக கருதுவது பயனுள்ளதாக இருக்கலாம். உண்மையில், pandas-இல் காணாமல் போன மதிப்புகளை நிர்வகிக்க நீங்கள் பயன்படுத்தும் முக்கியமான சில முறைமைகள் இந்த கருத்தை அவர்களின் பெயர்களில் பிரதிபலிக்கின்றன:

- `isnull()`: காணாமல் போன மதிப்புகளை குறிக்கும் Boolean மாஸ்கை உருவாக்குகிறது
- `notnull()`: `isnull()`-க்கு எதிர்மாறானது
- `dropna()`: தரவின் வடிகட்டப்பட்ட பதிப்பை திருப்பி வழங்குகிறது
- `fillna()`: காணாமல் போன மதிப்புகளை நிரப்பிய அல்லது மதிப்பீடு செய்த தரவின் நகலை திருப்பி வழங்குகிறது

இந்த முறைமைகள் மிகவும் முக்கியமானவை, மேலும் அவற்றை நன்கு புரிந்து கொண்டு பயன்படுத்துவதில் நிபுணத்துவம் பெறுவது அவசியம். எனவே, இவற்றை ஒவ்வொன்றாக விரிவாகப் பார்ப்போம்.


### வெற்றிட மதிப்புகளை கண்டறிதல்

இப்போது வெற்றிட மதிப்புகளின் முக்கியத்துவத்தை நாங்கள் புரிந்துகொண்டுள்ளோம், அவற்றை சமாளிப்பதற்கு முன், அவற்றை எங்கள் தரவுத்தொகுப்பில் கண்டறிய வேண்டும். 
`isnull()` மற்றும் `notnull()` ஆகியவை வெற்றிட தரவுகளை கண்டறிய உங்கள் முதன்மை முறைகள் ஆகும். இரண்டும் உங்கள் தரவின் மீது Boolean முகமூடிகளை திருப்பி அனுப்பும்.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

வெளியீட்டை நன்றாக கவனமாக பாருங்கள். இதில் ஏதேனும் உங்களை ஆச்சரியப்படுத்துகிறதா? `0` என்பது கணித ரீதியில் null ஆக இருந்தாலும், இது ஒரு சரியான முழு எண் மற்றும் pandas இதை அப்படியே ஏற்றுக்கொள்கிறது. `''` என்பது கொஞ்சம் நுணுக்கமானது. இதை Section 1-ல் காலியான string மதிப்பை பிரதிநிதித்துவப்படுத்த பயன்படுத்தினாலும், pandas-க்கு இது null ஆக இல்லாமல் ஒரு string பொருளாகவே உள்ளது.

இப்போது, இதை மாற்றி, நீங்கள் நடைமுறையில் பயன்படுத்தும் விதத்தில் இந்த முறைகளை பயன்படுத்துவோம். Boolean masks-ஐ நேரடியாக ``Series`` அல்லது ``DataFrame`` index ஆக பயன்படுத்தலாம், இது தனித்த missing (அல்லது உள்ள) மதிப்புகளுடன் வேலை செய்யும்போது பயனுள்ளதாக இருக்கும்.

நாம் missing values-இன் மொத்த எண்ணிக்கையை அறிய விரும்பினால், `isnull()` முறை உருவாக்கும் mask-ஐப் பயன்படுத்தி ஒரு sum செய்யலாம்.


In [19]:
example3.isnull().sum()

2

### பயிற்சி:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**முக்கியக் குறிப்புகள்**: `isnull()` மற்றும் `notnull()` முறைகள் இரண்டும் DataFrames-ல் பயன்படுத்தும்போது ஒரே மாதிரியான முடிவுகளை உருவாக்குகின்றன: அவை முடிவுகளை மற்றும் அந்த முடிவுகளின் குறியீட்டை காட்டுகின்றன, இது உங்கள் தரவுடன் போராடும் போது உங்களுக்கு மிகுந்த உதவியாக இருக்கும்.


### காணாமல் போன தரவுகளை கையாளுதல்

> **கற்றல் இலக்கு:** இந்த துணைப்பகுதியின் முடிவில், DataFrames-இல் null மதிப்புகளை எப்போது மற்றும் எப்படி மாற்றுவது அல்லது நீக்குவது என்பதை நீங்கள் அறிந்திருக்க வேண்டும்.

Machine Learning மாதிரிகள் காணாமல் போன தரவுகளை நேரடியாக கையாள முடியாது. எனவே, தரவுகளை மாதிரியில் அனுப்புவதற்கு முன், இந்த காணாமல் போன மதிப்புகளை கையாள வேண்டும்.

காணாமல் போன தரவுகளை எப்படி கையாளுகிறோம் என்பதில் நுண்ணிய சமநிலைகள் உள்ளன, இது உங்கள் இறுதி பகுப்பாய்வு மற்றும் நிஜ உலக முடிவுகளை பாதிக்கக்கூடும்.

காணாமல் போன தரவுகளை கையாளுவதற்கான முக்கியமான இரண்டு முறைகள் உள்ளன:

1.   காணாமல் போன மதிப்பை கொண்ட வரிசையை நீக்குதல்
2.   காணாமல் போன மதிப்பை வேறு ஒரு மதிப்புடன் மாற்றுதல்

இந்த இரண்டு முறைகளையும் மற்றும் அவற்றின் நன்மைகள் மற்றும் குறைகளை விரிவாக விவாதிக்கப் போகிறோம்.


### வெற்றிட மதிப்புகளை நீக்குதல்

நாம் எங்கள் மாதிரிக்கு அனுப்பும் தரவின் அளவு அதன் செயல்திறனை நேரடியாக பாதிக்கிறது. வெற்றிட மதிப்புகளை நீக்குவது என்பது தரவுப் புள்ளிகளின் எண்ணிக்கையை குறைப்பதைக் குறிக்கிறது, இதனால் தரவுத்தொகுப்பின் அளவும் குறைகிறது. எனவே, தரவுத்தொகுப்பு மிகவும் பெரியதாக இருந்தால், வெற்றிட மதிப்புகளுடன் உள்ள வரிசைகளை நீக்குவது பரிந்துரைக்கப்படுகிறது.

மற்றொரு சூழ்நிலை என்னவென்றால், ஒரு குறிப்பிட்ட வரிசை அல்லது நெடுவரிசையில் நிறைய காணாமல் போன மதிப்புகள் இருக்கலாம். அப்போது, அவற்றை நீக்கலாம், ஏனெனில் அந்த வரிசை/நெடுவரிசைக்கு பெரும்பாலான தரவுகள் காணாமல் போனதால், அவை எங்கள் பகுப்பாய்வுக்கு அதிக மதிப்பைச் சேர்க்காது.

காணாமல் போன மதிப்புகளை அடையாளம் காண்பதற்கு அப்பால், pandas `Series` மற்றும் `DataFrame`களிலிருந்து வெற்றிட மதிப்புகளை நீக்க ஒரு வசதியான வழியை வழங்குகிறது. இதை செயல்பாட்டில் காண, `example3`க்கு திரும்புவோம். `DataFrame.dropna()` செயல்பாடு வெற்றிட மதிப்புகளுடன் உள்ள வரிசைகளை நீக்க உதவுகிறது.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

இந்தது `example3[example3.notnull()]` என்ற உங்களின் வெளியீட்டைப் போலவே இருக்க வேண்டும் என்பதை கவனிக்கவும். இங்கு வேறுபாடு என்னவென்றால், மாஸ்க் செய்யப்பட்ட மதிப்புகளை மட்டும் குறிக்காமல், `dropna` அந்தக் காணாமல் போன மதிப்புகளை `Series` `example3`-இல் இருந்து நீக்கியுள்ளது.

DataFrames இரண்டு பரிமாணங்களைக் கொண்டுள்ளதால், தரவுகளை நீக்குவதற்கு மேலும் விருப்பங்களை வழங்குகின்றன.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(பாண்டாஸ் இரண்டு களங்களை `NaN`களை ஏற்க floats ஆக மாற்றியது கவனித்தீர்களா?)

நீங்கள் ஒரு `DataFrame`-இல் இருந்து ஒரு மதிப்பை மட்டும் நீக்க முடியாது, எனவே முழு வரிசைகள் அல்லது களங்களை நீக்க வேண்டும். நீங்கள் என்ன செய்கிறீர்கள் என்பதற்கேற்ப, ஒன்று அல்லது மற்றொன்றை செய்ய விரும்பலாம், எனவே பாண்டாஸ் இரண்டிற்கும் விருப்பங்களை வழங்குகிறது. காரணம், தரவியல் அறிவியலில், களங்கள் பொதுவாக மாறிகள் மற்றும் வரிசைகள் கண்காணிப்புகளை பிரதிநிதித்துவப்படுத்துகின்றன, நீங்கள் தரவின் வரிசைகளை நீக்க அதிகமாக வாய்ப்பு உள்ளது; `dropna()`-இன் இயல்பான அமைப்பு எந்த null மதிப்புகளையும் கொண்டுள்ள அனைத்து வரிசைகளையும் நீக்குவதற்காக அமைக்கப்பட்டுள்ளது:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

தேவையானால், நீங்கள் நெடுவரிசைகளிலிருந்து NA மதிப்புகளை நீக்கலாம். இதற்கு `axis=1` ஐ பயன்படுத்தவும்:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

குறிப்பாக, இது சிறிய தரவுத்தொகுப்புகளில் நீங்கள் வைத்திருக்க விரும்பும் பல தரவுகளை நீக்கக்கூடும். நீங்கள் பல அல்லது அனைத்து null மதிப்புகளை கொண்ட வரிசைகள் அல்லது நெடுவரிசைகளை மட்டும் நீக்க விரும்பினால் என்ன செய்ய வேண்டும்? `dropna`-இல் `how` மற்றும் `thresh` அளவுருக்களை குறிப்பிடலாம்.

இயல்பாக, `how='any'` (நீங்கள் சுயமாக சரிபார்க்க விரும்பினால் அல்லது முறை எந்த அளவுருக்களை கொண்டுள்ளது என்பதைப் பார்க்க விரும்பினால், ஒரு குறியீட்டு செலில் `example4.dropna?` இயக்கவும்). மாற்றாக, `how='all'`-ஐ குறிப்பிடலாம், இதனால் அனைத்து null மதிப்புகளை கொண்ட வரிசைகள் அல்லது நெடுவரிசைகளை மட்டும் நீக்கலாம். இதை செயல்பாட்டில் காண, நமது `DataFrame` எடுத்துக்காட்டை விரிவாக்கி, அடுத்த பயிற்சியில் இதைப் பார்ப்போம்.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> முக்கிய குறிப்புகள்:  
1. தரவுத்தொகுப்பு பெரியதாக இருந்தால் மட்டுமே null மதிப்புகளை நீக்குவது நல்ல யோசனை.  
2. முழு வரிசைகள் அல்லது நெடுவரிசைகள் பெரும்பாலான தரவுகள் இல்லை என்றால் அவற்றை நீக்கலாம்.  
3. `DataFrame.dropna(axis=)` முறை null மதிப்புகளை நீக்க உதவுகிறது. `axis` аргументம் வரிசைகளை நீக்க வேண்டுமா அல்லது நெடுவரிசைகளை நீக்க வேண்டுமா என்பதை குறிப்பிடுகிறது.  
4. `how` аргументம் கூட பயன்படுத்தப்படலாம். இயல்பாக இது `any` ஆக அமைக்கப்பட்டுள்ளது. எனவே, எந்த null மதிப்புகளையும் கொண்டுள்ள வரிசைகள்/நெடுவரிசைகளை மட்டும் இது நீக்குகிறது. இதை `all` ஆக அமைக்கலாம், அதனால் அனைத்து மதிப்புகளும் null ஆக இருக்கும் வரிசைகள்/நெடுவரிசைகளை மட்டும் நாம் நீக்குவோம் என்று குறிப்பிடலாம்.  


### பயிற்சி:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


`thresh` அளவுரு உங்களுக்கு நுண்ணிய கட்டுப்பாட்டை வழங்குகிறது: ஒரு வரி அல்லது நெடுவரிசை வைத்திருக்க தேவையான *non-null* மதிப்புகளின் எண்ணிக்கையை நீங்கள் அமைக்கலாம்:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

இங்கு, முதல் மற்றும் கடைசி வரிகள் நீக்கப்பட்டுள்ளன, ஏனெனில் அவற்றில் இரண்டு non-null மதிப்புகள் மட்டுமே உள்ளன.


### வெற்றிட மதிப்புகளை நிரப்புதல்

சில நேரங்களில், வெற்றிட மதிப்புகளை சரியானதாக இருக்கக்கூடியவற்றால் நிரப்புவது பொருத்தமாக இருக்கலாம். வெற்றிட மதிப்புகளை நிரப்ப சில நுட்பங்கள் உள்ளன. முதலில், டொமைன் அறிவு (தொகுப்பின் அடிப்படையாக உள்ள பொருள் பற்றிய அறிவு) பயன்படுத்தி, எப்படியாவது வெற்றிட மதிப்புகளை அணுகுமுறை செய்யலாம்.

நீங்கள் `isnull` ஐ பயன்படுத்தி இதை நேரடியாக செய்யலாம், ஆனால் இது கடினமாக இருக்கலாம், குறிப்பாக நீங்கள் நிரப்ப வேண்டிய மதிப்புகள் அதிகமாக இருந்தால். தரவியல் அறிவியலில் இது மிகவும் பொதுவான பணியாக இருப்பதால், pandas `fillna` ஐ வழங்குகிறது, இது உங்கள் தேர்வின் மூலம் வெற்றிட மதிப்புகளை மாற்றியமைக்கப்பட்ட `Series` அல்லது `DataFrame` இன் ஒரு நகலை திருப்புகிறது. இது நடைமுறையில் எப்படி செயல்படுகிறது என்பதை பார்க்க மற்றொரு உதாரண `Series` ஐ உருவாக்குவோம்.


### வகை தரவுகள் (எண் சார்ந்தவை அல்ல)
முதலில், எண் சார்ந்தவை அல்லாத தரவுகளைப் பார்ப்போம். தரவுத்தொகுப்புகளில், வகை தரவுகளைக் கொண்ட நெடுவரிசைகள் இருக்கும். உதாரணமாக, பாலினம், உண்மை அல்லது தவறு போன்றவை.

இவற்றில் பெரும்பாலான சந்தர்ப்பங்களில், நாங்கள் குறைவான மதிப்புகளை அந்த நெடுவரிசையின் `mode` மூலம் மாற்றுகிறோம். உதாரணமாக, 100 தரவுப் புள்ளிகள் உள்ளன, அதில் 90 பேர் "உண்மை" என்று கூறியுள்ளனர், 8 பேர் "தவறு" என்று கூறியுள்ளனர், மேலும் 2 பேர் எந்தவிதமான தகவலையும் நிரப்பவில்லை. அப்போது, முழு நெடுவரிசையைப் பொருத்து, அந்த 2 இடங்களை "உண்மை" என நிரப்பலாம்.

இங்கு மீண்டும், துறையியல் அறிவைப் பயன்படுத்தலாம். `mode` மூலம் நிரப்புவதற்கான ஒரு உதாரணத்தைப் பார்ப்போம்.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

இப்போது, முதலில் `None` மதிப்பை mode மூலம் நிரப்புவதற்கு முன் mode ஐ கண்டுபிடிப்போம்.


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

அதனால், None ஐ True ஆக மாற்றுவோம்


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

நாம் காணக்கூடியதுபோல், null மதிப்பு மாற்றப்பட்டுள்ளது. 'True' என்பதற்குப் பதிலாக எதையும் எழுதியிருக்கலாம், அது மாற்றப்பட்டிருக்கும்.


### எண் தரவுகள்
இப்போது, எண் தரவுகளுக்கு வருகிறோம். இங்கே, காணாமல் போன மதிப்புகளை மாற்றுவதற்கான இரண்டு பொதுவான முறைகள் உள்ளன:

1. வரிசையின் நடுக்கோட்டுடன் மாற்றவும்  
2. வரிசையின் சராசரியுடன் மாற்றவும்  

தரவு சாய்வு மற்றும் வெளிப்படையான மதிப்புகள் உள்ளபோது, நடுக்கோட்டுடன் மாற்றுகிறோம். காரணம், நடுக்கோடு வெளிப்படையான மதிப்புகளுக்கு எதிராக உறுதியாக இருக்கும்.

தரவு சீராக்கப்பட்டபோது, சராசரியை பயன்படுத்தலாம், ஏனெனில் அந்த நேரத்தில் சராசரி மற்றும் நடுக்கோடு மிகவும் அருகில் இருக்கும்.

முதலில், சாதாரணமாகப் பகிர்ந்தளிக்கப்பட்ட ஒரு நெடுவரிசையை எடுத்துக்கொள்வோம், மற்றும் அந்த நெடுவரிசையின் சராசரியுடன் காணாமல் போன மதிப்பை நிரப்புவோம்.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

கோட்டின் சராசரி என்பது


In [33]:
np.mean(fill_with_mean[0])

0.0

சராசையுடன் நிரப்புதல்


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

நாம் காணக்கூடியதுபோல், காணாமல் போன மதிப்பு அதன் சராசரியால் மாற்றப்பட்டுள்ளது.


இப்போது மற்றொரு தரவுத்தொகுப்பை முயற்சிப்போம், இந்த முறை None மதிப்புகளை அந்த நெடுவரிசையின் நடுக்கோட்டத்தால் மாற்றுவோம்.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

இரண்டாவது நெடுவரிசையின் நடுப்புள்ளி என்பது


In [36]:
fill_with_median[1].median()

4.0

மத்திய மதிப்பால் நிரப்புதல்


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

நாம் காணக்கூடியதுபோல், NaN மதிப்பு அந்த நெடுவரிசையின் நடுக்கணக்கால் மாற்றப்பட்டுள்ளது.


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

நீங்கள் அனைத்து காலியாக உள்ள பதிவுகளையும் `0` போன்ற ஒரு மதிப்புடன் நிரப்பலாம்:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> முக்கிய குறிப்புகள்:
1. பிழையான மதிப்புகளை நிரப்புவது குறைந்த தரவுகள் உள்ளபோது அல்லது பிழையான தரவுகளை நிரப்ப ஒரு உத்தி இருக்கும் போது செய்ய வேண்டும்.
2. பிழையான மதிப்புகளை அணுகுமுறையாக நிரப்ப துறை அறிவை பயன்படுத்தலாம்.
3. வகைப்பாட்டுத் தரவுகளுக்கு, பெரும்பாலும், பிழையான மதிப்புகள் அந்த நெடுவரிசையின் முறை (mode) மூலம் மாற்றப்படுகின்றன.
4. எண் தரவுகளுக்கு, பிழையான மதிப்புகள் பொதுவாக நெடுவரிசைகளின் சராசரி (சீரமைக்கப்பட்ட தரவுக்கட்டங்களுக்காக) அல்லது நடுக்கோட்டுடன் நிரப்பப்படுகின்றன.


### பயிற்சி:


In [40]:
# What happens if you try to fill null values with a string, like ''?


நீங்கள் **முன்னே-fill** செய்யும் null மதிப்புகளை நிரப்பலாம், அதாவது கடைசி செல்லுபடியாகும் மதிப்பை null நிரப்ப பயன்படுத்தலாம்:


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

நீங்கள் அடுத்த செல்லுபடியாகும் மதிப்பை பின்புறமாக பரப்ப **பின்நிரப்ப**வும் செய்து ஒரு வெற்றிடத்தை நிரப்பலாம்:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

நீங்கள் ஊகிக்கலாம் போல, இது DataFrames உடன் ஒரே மாதிரியான முறையில் செயல்படுகிறது, ஆனால் நீங்கள் null மதிப்புகளை நிரப்ப ஒரு `axis` ஐ குறிப்பிடவும் முடியும்:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

முன்னைய மதிப்பு முன்னோக்கி நிரப்புவதற்கு கிடைக்காதபோது, வெற்றிட மதிப்பு மாறாமல் இருக்கும் என்பதை கவனிக்கவும்.


### பயிற்சி:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


நீங்கள் `fillna` பயன்படுத்துவதில் சிருஷ்டியாக இருக்கலாம். உதாரணமாக, `example4` ஐ மீண்டும் பார்ப்போம், ஆனால் இந்த முறை `DataFrame` இல் உள்ள அனைத்து மதிப்புகளின் சராசரியுடன் காணாமல் போன மதிப்புகளை நிரப்புவோம்:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

குறிப்பாக, மூன்றாவது நெடுவரிசை இன்னும் மதிப்பில்லாமல் உள்ளது: இயல்பான திசை மதிப்புகளை வரிசை வாரியாக நிரப்புவது.

> **முக்கியக் கருத்து:** உங்கள் தரவுத்தொகுப்புகளில் காணப்படும் பிழைமதிப்புகளை சமாளிக்க பல வழிகள் உள்ளன. நீங்கள் பயன்படுத்தும் குறிப்பிட்ட உத்தி (அவற்றை நீக்குவது, மாற்றுவது, அல்லது எப்படி மாற்றுவது) அந்த தரவின் தன்மைகளால் தீர்மானிக்கப்பட வேண்டும். நீங்கள் தரவுத்தொகுப்புகளை அதிகமாக கையாளும் போது, பிழைமதிப்புகளை எப்படி சமாளிக்க வேண்டும் என்பதில் சிறந்த புரிதலை உருவாக்குவீர்கள்.


### வகைமுறை தரவுகளை குறியாக்கம் செய்வது

இயந்திரக் கற்றல் மாதிரிகள் எண் மற்றும் எண் வடிவிலான தரவுகளுடன் மட்டுமே செயல்பட முடியும். அவை "ஆம்" மற்றும் "இல்லை" என்பவற்றின் வித்தியாசத்தை புரிந்துகொள்ள முடியாது, ஆனால் 0 மற்றும் 1 என்பவற்றை வேறுபடுத்த முடியும். எனவே, காணாமல் போன மதிப்புகளை நிரப்பிய பிறகு, மாதிரி புரிந்துகொள்ள வகைமுறை தரவுகளை எண் வடிவமாக குறியாக்கம் செய்ய வேண்டும்.

குறியாக்கம் இரண்டு வழிகளில் செய்யலாம். அடுத்ததாக அவற்றைப் பற்றி விவாதிக்கப் போகிறோம்.


**லேபல் என்கோடிங்**

லேபல் என்கோடிங் என்பது ஒவ்வொரு வகையையும் ஒரு எண்ணாக மாற்றுவது ஆகும். உதாரணமாக, நாம் விமான பயணிகளின் தரவுத்தொகுப்பை எடுத்துக்கொண்டால், அதில் ['பிஸினஸ் கிளாஸ்', 'எகானமி கிளாஸ்', 'ஃபர்ஸ்ட் கிளாஸ்'] ஆகியவற்றில் இருந்து பயண வகையை கொண்ட ஒரு பத்தி இருக்கும். இதற்கு லேபல் என்கோடிங் செய்யப்படும் போது, இது [0,1,2] ஆக மாற்றப்படும். இதை ஒரு உதாரணமாகக் கொண்டு கோடில் பார்க்கலாம். எதிர்கால நோட்புக்குகளில் `scikit-learn` பற்றி கற்றுக்கொள்ள இருப்பதால், இங்கு அதை பயன்படுத்தமாட்டோம்.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

முதல் sütunில் லேபல் குறியாக்கத்தை செய்ய, ஒவ்வொரு வகுப்பிற்கும் ஒரு எண்ணை ஒதுக்குவதற்கான வரைபடத்தை முதலில் வரையறுக்க வேண்டும், பின்னர் மாற்ற வேண்டும்.


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

எப்படி நாம் எதிர்பார்த்தோமோ, அவ்வாறே வெளியீடு பொருந்துகிறது. எனவே, எப்போது லேபல் என்கோடிங்கைப் பயன்படுத்த வேண்டும்? லேபல் என்கோடிங்கை பின்வரும் நிலைகளில் பயன்படுத்தலாம்:  
1. வகைகளின் எண்ணிக்கை அதிகமாக இருக்கும் போது  
2. வகைகள் வரிசையில் இருக்கும் போது  


**ஒன் ஹாட் என்கோடிங்**

ஒன் ஹாட் என்கோடிங் என்பது மற்றொரு வகையான என்கோடிங் முறையாகும். இந்த என்கோடிங் முறையில், ஒவ்வொரு பாகத்தின் வகையும் தனி பாகமாக சேர்க்கப்படும், மேலும் ஒவ்வொரு தரவுப் புள்ளியும் அந்த வகையை கொண்டிருக்கிறதா என்பதைப் பொறுத்து 0 அல்லது 1 பெறும். எனவே, n வெவ்வேறு வகைகள் இருந்தால், n பாகங்கள் டேட்டாபிரேமில் சேர்க்கப்படும்.

உதாரணமாக, அதே விமான வகுப்பின் உதாரணத்தை எடுத்துக்கொள்வோம். வகைகள்: ['பிஸினஸ் கிளாஸ்', 'எகனாமி கிளாஸ்', 'பர்ஸ்ட் கிளாஸ்'] . எனவே, ஒன் ஹாட் என்கோடிங் செய்யும்போது, பின்வரும் மூன்று பாகங்கள் தரவுத்தொகுப்பில் சேர்க்கப்படும்: ['class_business class', 'class_economy class', 'class_first class'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

முதல் நெடுவரிசையில் ஒரே-தீவிர குறியாக்கத்தை செய்யலாம்.


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

ஒவ்வொரு ஹாட் என்கோடு செய்யப்பட்ட நெடுவம் 0 அல்லது 1 ஐ கொண்டுள்ளது, இது அந்த தரவுப் புள்ளிக்கான அந்த வகை உள்ளது என்பதை குறிப்பிடுகிறது.


ஒரு ஹாட் என்கோடிங்கை எப்போது பயன்படுத்த வேண்டும்? ஒரு ஹாட் என்கோடிங்கை பின்வரும் நிலைகளில் ஒன்று அல்லது இரண்டிலும் பயன்படுத்தலாம்:

1. வகைகளின் எண்ணிக்கையும் தரவுத்தொகுப்பின் அளவும் குறைவாக இருக்கும் போது.  
2. வகைகள் எந்தவொரு குறிப்பிட்ட வரிசையையும் பின்பற்றாத போது.  


> முக்கிய குறிப்புகள்:
1. எண்களல்லாத தரவுகளை எண் தரவாக மாற்றுவதற்காக குறியாக்கம் செய்யப்படுகிறது.
2. குறியாக்கம் இரண்டு வகைகளாக உள்ளது: லேபல் குறியாக்கம் மற்றும் ஒன் ஹாட் குறியாக்கம், இவை dataset-இன் தேவைகளின் அடிப்படையில் செய்யப்படலாம்.


## மறுபடியும் வரும் தரவுகளை நீக்குதல்

> **கற்கும் இலக்கு:** இந்த துணைபகுதியின் முடிவில், DataFrames-இல் மறுபடியும் வரும் மதிப்புகளை அடையாளம் காணவும், அவற்றை நீக்கவும் நீங்கள் வசதியாக இருக்க வேண்டும்.

காணாமல் போன தரவுகளுக்கு கூடுதலாக, நீங்கள் உண்மையான உலக தரவுத்தொகுப்புகளில் அடிக்கடி மறுபடியும் வரும் தரவுகளை சந்திக்கலாம். அதிர்ஷ்டவசமாக, pandas மறுபடியும் வரும் பதிவுகளை கண்டறியவும், அவற்றை நீக்கவும் எளிய வழிமுறையை வழங்குகிறது.


### நகல் மதிப்புகளை கண்டறிதல்: `duplicated`

நீங்கள் pandas இல் உள்ள `duplicated` முறைமையைப் பயன்படுத்தி நகல் மதிப்புகளை எளிதாக கண்டறியலாம், இது ஒரு Boolean mask ஐ திருப்புகிறது, அதில் ஒரு `DataFrame` இல் உள்ள ஒரு பதிவே முன்பே உள்ளதற்கான நகலா என்பதை காட்டுகிறது. இதை செயல்பாட்டில் காண மற்றொரு உதாரண `DataFrame` ஐ உருவாக்கலாம்.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### நகல்களை நீக்குதல்: `drop_duplicates`
`drop_duplicates` என்பது `duplicated` மதிப்புகள் அனைத்தும் `False` ஆக இருக்கும் தரவின் ஒரு நகலை திருப்பி அளிக்கிறது:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

`duplicated` மற்றும் `drop_duplicates` இரண்டும் இயல்பாக அனைத்து நெடுவரிசைகளையும் கருத்தில் கொள்கின்றன, ஆனால் உங்கள் `DataFrame`-இல் அவை ஒரு பகுதி நெடுவரிசைகளை மட்டுமே பரிசீலிக்க வேண்டும் என்று நீங்கள் குறிப்பிடலாம்:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **குறிப்பு:** நகல் தரவுகளை நீக்குவது ஒவ்வொரு தரவியல் அறிவியல் திட்டத்திலும் முக்கியமான பகுதியாகும். நகல் தரவுகள் உங்கள் பகுப்பாய்வுகளின் முடிவுகளை மாற்றி தவறான முடிவுகளை வழங்கக்கூடும்!


## உண்மையான தரவின் தரச்செயல்முறைகள்

> **கற்கும் நோக்கம்:** இந்த பிரிவின் முடிவில், நீங்கள் பொதுவான உண்மையான தரத்தின் தரச்சிக்கல்களை கண்டறிந்து சரிசெய்யும் திறனை பெறுவீர்கள், இதில் முரண்பட்ட வகைமுறை மதிப்புகள், அசாதாரண எண் மதிப்புகள் (விலகல்கள்), மற்றும் சிறிய மாற்றங்களுடன் கூடிய நகல் நிறுவனங்கள் அடங்கும்.

இல்லாத மதிப்புகள் மற்றும் துல்லியமான நகல்கள் பொதுவான சிக்கல்களாக இருந்தாலும், உண்மையான தரவுத்தொகுப்புகளில் மேலும் நுணுக்கமான பிரச்சினைகள் காணப்படுகின்றன:

1. **முரண்பட்ட வகைமுறை மதிப்புகள்**: ஒரே வகை வெவ்வேறு முறையில் எழுதப்பட்டிருக்கும் (எ.கா., "USA", "U.S.A", "United States")
2. **அசாதாரண எண் மதிப்புகள்**: தரவின் உள்ளீட்டுப் பிழைகளை சுட்டிக்காட்டும் மிகுந்த விலகல்கள் (எ.கா., வயது = 999)
3. **சமீபத்திய நகல் வரிசைகள்**: சிறிய மாற்றங்களுடன் ஒரே நிறுவனத்தை பிரதிநிதித்துவம் செய்யும் பதிவுகள்

இந்த சிக்கல்களை கண்டறிந்து கையாளும் நுட்பங்களை ஆராய்வோம்.


### "கழுவாத" மாதிரி தரவுத்தொகுப்பை உருவாக்குதல்

முதலில், நிஜ உலக தரவுகளில் நாம் அடிக்கடி சந்திக்கும் பிரச்சனைகள் கொண்ட ஒரு மாதிரி தரவுத்தொகுப்பை உருவாக்குவோம்:


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. ஒத்தில்லாத வகைமுறை மதிப்புகளை கண்டறிதல்

`country` நெடுவரிசையில் ஒரே நாடுகளுக்கான பல்வேறு பிரதிநிதித்துவங்கள் உள்ளன. இந்த ஒத்தில்லாமைகளை கண்டறிவோம்:


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### வகைமுறை மதிப்புகளை ஒரே மாதமாக்குதல்

இந்த மதிப்புகளை ஒரே மாதமாக்க ஒரு மாப்பிங் உருவாக்கலாம். ஒரு எளிய முறையாக, lowercase-க்கு மாற்றி, ஒரு மாப்பிங் dictionary உருவாக்கலாம்:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**மாற்று வழி: மங்கலான பொருத்தத்தை பயன்படுத்துதல்**

மிகவும் சிக்கலான நிலைகளுக்கு, `rapidfuzz` நூலகத்தை பயன்படுத்தி மங்கலான சரம் பொருத்தத்தை கொண்டு ஒரே மாதிரியான சரங்களை தானாக கண்டறியலாம்:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. எண் மதிப்புகளில் அசாதாரணங்களை (வெளியிலிகள்) கண்டறிதல்

`age` பத்தியைப் பார்க்கும்போது, 199 மற்றும் -5 போன்ற சந்தேகமான மதிப்புகள் உள்ளன. இந்த வெளியிலிகளை கண்டறிய புள்ளியியல் முறைகளைப் பயன்படுத்தலாம்.


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### IQR (இடைக்கால வரம்பு) முறையைப் பயன்படுத்துதல்

IQR முறை என்பது மிகுந்த மதிப்புகளுக்கு குறைவாக பாதிக்கப்படும் ஒரு வலுவான புள்ளியியல் தொழில்நுட்பமாகும், இது வெளிப்படையான மதிப்புகளை கண்டறிய உதவுகிறது:


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### Z-Score முறையைப் பயன்படுத்துதல்

Z-Score முறை சராசரியிலிருந்து நிலையான சிதறல்களை அடிப்படையாகக் கொண்டு வெளிப்படையான மதிப்புகளை அடையாளம் காண்கிறது:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### அவதானமான மதிப்புகளை கையாளுதல்

அவதானமான மதிப்புகள் கண்டறியப்பட்ட பிறகு, அவற்றை பல வழிகளில் கையாளலாம்:
1. **நீக்குதல்**: தவறானவை என்றால் அவதானமான மதிப்புகளைக் கொண்ட வரிசைகளை நீக்கவும்
2. **வரம்பு அமைத்தல்**: எல்லை மதிப்புகளால் மாற்றவும்
3. **NaN-ஆக மாற்றுதல்**: இவை காணாமல் போன தரவாக கருதி நிரப்பும் முறைகளைப் பயன்படுத்தவும்
4. **வைத்திருக்கவும்**: அவை உண்மையான மிகுந்த மதிப்புகள் என்றால்


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. அருகிலுள்ள நகல் வரிகளை கண்டறிதல்

எங்கள் தரவுத்தொகுப்பில் "John Smith" என்பவருக்கான பல பதிவுகள் உள்ளன, அவற்றில் சிலவற்றின் மதிப்புகள் சிறிது மாறுபட்டுள்ளன. பெயர் ஒற்றுமையை அடிப்படையாகக் கொண்டு சாத்தியமான நகல்களை கண்டறிவோம்.


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### பனுவல் பொருத்தத்துடன் அருகிலுள்ள நகல்களை கண்டறிதல்

மேம்பட்ட நகல் கண்டறிதலுக்காக, நாம் பனுவல் பொருத்தத்தை பயன்படுத்தி ஒரே மாதிரியான பெயர்களை கண்டறியலாம்:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### நகல் தரவுகளை கையாளுதல்

ஒருமுறை கண்டறிந்த பின், நகல் தரவுகளை எப்படி கையாள வேண்டும் என்பதை தீர்மானிக்க வேண்டும்:
1. **முதல் நிகழ்வை வைத்திருக்கவும்**: `drop_duplicates(keep='first')` பயன்படுத்தவும்
2. **கடைசி நிகழ்வை வைத்திருக்கவும்**: `drop_duplicates(keep='last')` பயன்படுத்தவும்
3. **தகவலை ஒருங்கிணைக்கவும்**: நகல் வரிகளிலிருந்து தகவலை இணைக்கவும்
4. **கையேடு மதிப்பீடு**: மனித மதிப்பீட்டிற்காக குறிக்கவும்


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### சுருக்கம்: முழுமையான தரவுத் தூய்மைப்படுத்தும் குழாய்

இதை முழுமையான தூய்மைப்படுத்தும் குழாயாக ஒன்றிணைக்கலாம்:


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 சவால் பயிற்சி

இப்போது உங்கள் முறை! கீழே தரவின் புதிய வரி கொடுக்கப்பட்டுள்ளது, இதில் பல தரம் குறைவான பிரச்சினைகள் உள்ளன. நீங்கள் செய்ய வேண்டியது:

1. இந்த வரியில் உள்ள அனைத்து பிரச்சினைகளையும் கண்டறியவும்  
2. ஒவ்வொரு பிரச்சினையையும் சுத்தமாக்குவதற்கான குறியீட்டை எழுதவும்  
3. சுத்தமாக்கப்பட்ட வரியை தரவுத்தொகுப்பில் சேர்க்கவும்  

இங்கே பிரச்சினையுள்ள தரவுகள்:


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### முக்கிய குறிப்புகள்

1. **மாறுபட்ட வகைகள்** உண்மையான தரவுகளில் பொதுவாக காணப்படும். தனித்துவமான மதிப்புகளை சரிபார்த்து, மாப்பிங் அல்லது மங்கலான பொருத்தம் மூலம் அவற்றை ஒரே மாதமாக்கவும்.

2. **விலகிய மதிப்புகள்** உங்கள் பகுப்பாய்வில் பெரிதும் பாதிப்பை ஏற்படுத்தலாம். துறையின் அறிவு மற்றும் புள்ளியியல் முறைகளை (IQR, Z-score) இணைத்து அவற்றை கண்டறியவும்.

3. **சமீபத்திய நகல்கள்** சரியான நகல்களை விட கண்டறிய கடினமாக இருக்கும். மங்கலான பொருத்தம் மற்றும் தரவுகளை சாதாரணமாக்கல் (எழுத்துகளை சிறிய எழுத்தாக மாற்றுதல், வெற்றிடங்களை நீக்குதல்) மூலம் அவற்றை அடையாளம் காண முயற்சிக்கவும்.

4. **தரவு சுத்திகரிப்பு மீண்டும் மீண்டும் செய்யப்படும் செயலாகும்**. பல முறைகளை பயன்படுத்தி, முடிவுகளை மதிப்பீடு செய்து, சுத்திகரிக்கப்பட்ட தரவுத்தொகுப்பை இறுதியாக முடிக்க வேண்டும்.

5. **உங்கள் முடிவுகளை பதிவு செய்யவும்**. நீங்கள் எந்த சுத்திகரிப்பு நடவடிக்கைகளை எடுத்தீர்கள் மற்றும் ஏன் எடுத்தீர்கள் என்பதை பதிவு செய்யுங்கள், இது மீண்டும் செய்யக்கூடியதற்கும் வெளிப்படைத்தன்மைக்கும் முக்கியமானது.

> **சிறந்த நடைமுறை:** உங்கள் அசல் "அழுக்கான" தரவின் ஒரு பிரதியை எப்போதும் வைத்திருங்கள். உங்கள் மூல தரவுக் கோப்புகளை ஒருபோதும் மீட்டெடுக்க வேண்டாம் - `data_cleaned.csv` போன்ற தெளிவான பெயரிடல் முறைகளுடன் சுத்திகரிக்கப்பட்ட பதிப்புகளை உருவாக்கவும்.



---

**குறிப்பு**:  
இந்த ஆவணம் [Co-op Translator](https://github.com/Azure/co-op-translator) என்ற AI மொழிபெயர்ப்பு சேவையைப் பயன்படுத்தி மொழிபெயர்க்கப்பட்டுள்ளது. நாங்கள் துல்லியத்திற்காக முயற்சிக்கின்றோம், ஆனால் தானியக்க மொழிபெயர்ப்புகளில் பிழைகள் அல்லது தவறான தகவல்கள் இருக்கக்கூடும் என்பதை தயவுசெய்து கவனத்தில் கொள்ளவும். அதன் தாய்மொழியில் உள்ள மூல ஆவணம் அதிகாரப்பூர்வ ஆதாரமாக கருதப்பட வேண்டும். முக்கியமான தகவல்களுக்கு, தொழில்முறை மனித மொழிபெயர்ப்பு பரிந்துரைக்கப்படுகிறது. இந்த மொழிபெயர்ப்பைப் பயன்படுத்துவதால் ஏற்படும் எந்த தவறான புரிதல்கள் அல்லது தவறான விளக்கங்களுக்கு நாங்கள் பொறுப்பல்ல.
